// apollo-engine/lib/__tests__/engine-restify.test.js
"use strict";
// We put the restify tests in their own file because importing restify breaks
// instances of express in the same process:
// https://github.com/restify/node-restify/issues/1540
Object.defineProperty(exports, "__esModule", { value: true });
var apollo_server_restify_1 = require("apollo-server-restify");
var restify = require("restify");
var schema_1 = require("./schema");
var engine_common_1 = require("./engine-common");
engine_common_1.runSuitesForHttpServerFramework('restify', {
    createApp: function () {
        var server = restify.createServer({ name: 'test server' });
        var graphQLOptions = {
            schema: schema_1.schema,
            rootValue: schema_1.rootValue,
            tracing: true,
            cacheControl: true,
        };
        server.use(restify.plugins.bodyParser());
        server.use(restify.plugins.queryParser());
        server.post('/graphql', apollo_server_restify_1.graphqlRestify(graphQLOptions));
        server.get('/graphql', apollo_server_restify_1.graphqlRestify(graphQLOptions));
        server.get('/graphql/ping', function (req, res, next) {
            res.send({ pong: true });
            next();
        });
        return server;
    },
    serverForApp: function (app) {
        return app.server;
    },
    appParameter: 'restifyServer',
});
