// apollo-engine/lib/__tests__/engine-node6.test.js
"use strict";
// We put the micro tests in this file because you can't even import hapi or
// koa-bodyparser on versions of Node older than 6 (due to destructuring
// assignment).
Object.defineProperty(exports, "__esModule", { value: true });
var apollo_server_micro_1 = require("apollo-server-micro");
var micro_1 = require("micro");
var microRouter = require("microrouter");
var schema_1 = require("./schema");
var engine_common_1 = require("./engine-common");
engine_common_1.runSuitesForHttpServerFramework('micro', {
    createApp: function () {
        var handler = apollo_server_micro_1.microGraphql({
            schema: schema_1.schema,
            rootValue: schema_1.rootValue,
            tracing: true,
            cacheControl: true,
        });
        return micro_1.default(microRouter.router(microRouter.get('/graphql/ping', function () {
            return JSON.stringify({ pong: true });
        }), microRouter.get('/graphql', handler), microRouter.get('/graphql/', handler), microRouter.post('/graphql', handler), microRouter.post('/graphql/', handler)));
    },
    serverForApp: function (app) {
        return app;
    },
    appParameter: 'httpServer',
});
